src/regexmanager.cpp

```diff
@@ -222,18 +223,57 @@ void RegexManager::remove_last_regex(const std::string& location)
 	regexes.erase(it);
 }
 
-std::string RegexManager::extract_initial_marker(const std::string& str)
+std::string RegexManager::extract_outer_marker(const std::string str, const int index)
 {
+	regmatch_t pmatch;
+	regex_t regex;
+	// Non-zero number or non-angle bracket characters, enclosed in angle brackets
+	const char *regexpat = "<[^<>]+>";
+	std::string close = "</>";
+	std::string tmptag;
+	std::stack<std::string> tagstack;
+	int offset = 0;
+
 	if (str.length() == 0)
 		return "";
+	// Compile regex
+	if ( regcomp(&regex, regexpat, REG_EXTENDED) != 0 )
+		return "";
 
-	if (str[0] == '<') {
-		std::string::size_type pos = str.find_first_of(">", 0);
-		if (pos != std::string::npos) {
-			return str.substr(0, pos + 1);
+	int err = regexec(&regex, str.c_str(), 1, &pmatch, 0);
+	while (err == 0) {
+		// Get found tag
+		tmptag = str.substr( pmatch.rm_so+offset, pmatch.rm_eo-pmatch.rm_so);
+		// If the found tag is after the spot we're looking for
+		if (pmatch.rm_so+offset > index ){
+			if(!tagstack.empty() ) {
+				return tagstack.top();
+			} else {
+				return "";
+			}
+		}
+		// If the found tag is a closeing tag
+		if (tmptag == close) {
+			//If a tag is closed without a partner error out
+			if (tagstack.empty()) {
+				return "";
+			} else {
+				//otherwise pop it off the stack
+				tagstack.pop();
+			}
+		} else { //othewise add the tag to the stack
+			tagstack.push(tmptag);
 		}
+		offset += pmatch.rm_eo;
+		err = regexec(&regex, str.c_str() + offset, 1, &pmatch, 0);
 	}
-	return "";
+
+	if(!tagstack.empty()) {
+		return tagstack.top();
+	}
+
+		return "";
```

Comment:
Please fix the indentation here.

Also, this function leaks memory because `regfree` is never called on `regex`. I understand that the rest of the code here uses C's <regex.h>, but I suggest to rewrite this with C++'s <regex>.